Userland/Libraries/LibWeb/HTML/HTMLFormElement.idl

```diff
@@ -16,6 +16,30 @@ enum EnctypeAttribute {
     "text/plain"
 };
 
+// https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#attr-fs-method
+[MissingValueDefault=get, InvalidValueDefault=get]
+enum MethodAttribute {
+    "get",
+    "post",
+    "dialog"
+};
+
+// https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#attr-fs-formenctype
+[InvalidValueDefault=application/x-www-form-urlencoded]
+enum FormEnctypeAttribute {
+    "application/x-www-form-urlencoded",
+    "multipart/form-data",
+    "text/plain"
+};
+
+// https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#attr-fs-formmethod
+[InvalidValueDefault=get]
+enum FormMethodAttribute {
+    "get",
+    "post",
+    "dialog"
+};
```

Comment:
The enums `FormMethodAttribute` and `FormEnctypeAttribute` have no `MissingValueDefault` as per spec, but the test results are still incorrect.

I looked at the generated code and it seems correct, if there is no value it returns an empty string:
```cpp
if (!contentAttributeValue.has_value())
    retval = ""_string;
```

I'm not sure of how to fix this problem

Reply:
You'll want to change the logic in `Meta/Lagom/Tools/CodeGenerators/LibWeb/BindingsGenerator/IDLGenerators.cpp` line 3494 and beyond.

The current implementation doesn't take into account the "no such value" case gracefully.

```
                        attribute_generator.set("missing_enum_default_value", missing_value_default.has_value() ? missing_value_default.value().view() : ""sv);
                        attribute_generator.set("invalid_enum_default_value", invalid_value_default.has_value() ? invalid_value_default.value().view() : ""sv);
```

Defaulting to the empty StringView here seems wrong, as we should instead return `undefined` or `null` or some such value as appropriate.

However, there is also this:

```
                    // 5. If contentAttributeValue is null, then return the empty string.
                    // 6. Return contentAttributeValue.
```

Which suggests that there could be some inconsistencies between <https://html.spec.whatwg.org/multipage/common-dom-interfaces.html#reflecting-content-attributes-in-idl-attributes> and <https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#keywords-and-enumerated-attributes> ?

Reply:
Actually yeah, why do you believe that the empty string is incorrect? The HTML spec says explicitly that we should use the empty string. Do other browsers do something different? If so, we should open a spec issue to HTML.

> 4. If attributeDefinition indicates it is an [enumerated attribute](https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#enumerated-attribute) and the [reflected IDL attribute](https://html.spec.whatwg.org/multipage/common-dom-interfaces.html#reflected-idl-attribute) is defined to be [limited to only known values](https://html.spec.whatwg.org/multipage/common-dom-interfaces.html#limited-to-only-known-values):
>      1. If contentAttributeValue does not correspond to any state of attributeDefinition (e.g., it is null and there is no [missing value default](https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#missing-value-default)), or if it is in a state of attributeDefinition with no associated keyword value, then return the empty string.
>     2. Return the [canonical keyword](https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#canonical-keyword) for the state of attributeDefinition that contentAttributeValue corresponds to.
> 5. If contentAttributeValue is null, then return the empty string.
> 6. Return contentAttributeValue.

Reply:
Oh I misunderstood, "" is indeed correct but the test results are giving us 'get' and such. That does seem wrong, and is likely a bug in that part of the IDL Generator :sweat_smile:

Reply:
> Oh I misunderstood, "" is indeed correct but the test results are giving us 'get' and such. That does seem wrong, and is likely a bug in that part of the IDL Generator 😅

Yes "" is the correct value, to my untrained eye the generated code for the <button> formMethod getter seems correct, but it doesn't do what it should
Here the generated code (I added a comment in the relevant part)
```cpp
JS_DEFINE_NATIVE_FUNCTION(HTMLButtonElementPrototype::form_method_getter)
{
    WebIDL::log_trace(vm, "HTMLButtonElementPrototype::form_method_getter");
    [[maybe_unused]] auto& realm = *vm.current_realm();
    [[maybe_unused]] auto* impl = TRY(impl_from(vm));

    auto& relevant_agent = HTML::relevant_agent(*impl);
    auto* custom_data = verify_cast<Bindings::WebEngineCustomData>(relevant_agent.custom_data());
    auto& reactions_stack = custom_data->custom_element_reactions_stack;
    reactions_stack.element_queue_stack.append({});

    auto contentAttributeValue = impl->attribute(HTML::AttributeNames::formmethod);

    auto retval = contentAttributeValue.value_or(String {});

    // Here it checks if there is no value and returns an empty string, as it should
    // But why in the test "get" is returned
    if (!contentAttributeValue.has_value())
        retval = ""_string;

    Array valid_values { "get"_string, "post"_string, "dialog"_string };

    auto found = false;
    for (auto const& value : valid_values) {
        if (value.equals_ignoring_ascii_case(retval)) {
            found = true;
            retval = value;
            break;
        }
    }

    if (!found)
        retval = "get"_string;
    
    auto queue = reactions_stack.element_queue_stack.take_last();
    Bindings::invoke_custom_element_reactions(queue);

    return JS::PrimitiveString::create(vm, retval);

}
```

Reply:
> ```c++
> JS_DEFINE_NATIVE_FUNCTION(HTMLButtonElementPrototype::form_method_getter)
> {
>     WebIDL::log_trace(vm, "HTMLButtonElementPrototype::form_method_getter");
>     [[maybe_unused]] auto& realm = *vm.current_realm();
>     [[maybe_unused]] auto* impl = TRY(impl_from(vm));
> 
>     auto& relevant_agent = HTML::relevant_agent(*impl);
>     auto* custom_data = verify_cast<Bindings::WebEngineCustomData>(relevant_agent.custom_data());
>     auto& reactions_stack = custom_data->custom_element_reactions_stack;
>     reactions_stack.element_queue_stack.append({});
> 
>     auto contentAttributeValue = impl->attribute(HTML::AttributeNames::formmethod);
> 
>     auto retval = contentAttributeValue.value_or(String {});
> 
>     // Here it checks if there is no value and returns an empty string, as it should
>     // But why in the test "get" is returned
>     if (!contentAttributeValue.has_value())
>         retval = ""_string;
> 
>     Array valid_values { "get"_string, "post"_string, "dialog"_string };
> 
>     auto found = false;
>     for (auto const& value : valid_values) {
>         if (value.equals_ignoring_ascii_case(retval)) {
>             found = true;
>             retval = value;
>             break;
>         }
>     }
>     // this is the place where you return get instead
>     if (!found)
>         retval = "get"_string;
>     
>     auto queue = reactions_stack.element_queue_stack.take_last();
>     Bindings::invoke_custom_element_reactions(queue);
> 
>     return JS::PrimitiveString::create(vm, retval);
> 
> }
> ```

Since the empty string doesn't match any of the possible values value of found is never changed to true, meaning you hit the if(!found) case, where it does in fact change the value from an empty string to "get"

Reply:
Looks like the empty string is not a valid value, so its not allowed to be returned. see this part of the generated code
```
    if (!found)
        retval = "get"_string;
```

Reply:
Now that I look at the generated code again, I notice that `retval` is always overwritten by the later `if(!found)`
```cpp
JS_DEFINE_NATIVE_FUNCTION(HTMLButtonElementPrototype::form_method_getter)
{
    // ...snip...

    // Here we set retval to ""
    if (!contentAttributeValue.has_value())
        retval = ""_string;

    Array valid_values { "get"_string, "post"_string, "dialog"_string };

    // If there is no value this won't find any of the valid_values
    auto found = false;
    for (auto const& value : valid_values) {
        if (value.equals_ignoring_ascii_case(retval)) {
            found = true;
            retval = value;
            break;
        }
    }

    // So this if will be executed and retval will be overwritten
    if (!found)
        retval = "get"_string;
    
    // ...snip...
}
```

So yeah it's probably a bug

Reply:
Yea this is a bug in the IDL generator

Reply:
An easy way to make this behave closer to spec would be to add `""` to the `FormEnytype` enum in `HTMLFormElement.idl`. 

A better way to solve the problem would be to modify the way the IDL code generator handles the `MissingValueDefault` extended attribute to ensure he missing value default is included in the set of valid values (see: [here](https://github.com/LadybirdBrowser/ladybird/blob/97ca6036fa5dc1e73b27115304d8d435b9fc7df6/Meta/Lagom/Tools/CodeGenerators/LibWeb/BindingsGenerator/IDLGenerators.cpp#L3491-L3496) and [here](https://github.com/LadybirdBrowser/ladybird/blob/97ca6036fa5dc1e73b27115304d8d435b9fc7df6/Meta/Lagom/Tools/CodeGenerators/LibWeb/BindingsGenerator/IDLGenerators.cpp#L3548-L3556)

Reply:
> An easy way to make this behave closer to spec would be to add `""` to the `FormEnytype` enum in `HTMLFormElement.idl`.

This will fix that testcase but will break another. If the attribute is set to "" the value should be replaced with the `InvalidValueDefault` and not be accepted as a valid value